### PR TITLE
New API and static analysis

### PR DESCRIPTION
**Pull request by pavel-kirienko**
**Please focus on the documentation in `udpard.h`.** The other changes are trivial and can be ignored.

Changes:

1. Update the API according to the new design principle that the library is to be used with a third-party UDP/IP+IGMP stack.

> Q: Could you be a bit more specific about the multiplexing that you saw in the old version? Do you mean redundant interfaces?
> A: No, I mean the socket layer (or the lack-of-socket layer, as Erik puts it) already demultiplexes incoming datagrams per subject. The original design joins the demultiplexed datagram feeds into one feed which is then demultiplexed back through AVL tree lookup. This is suboptimal. The new design eliminates one (of the two) AVL lookups from the RX pipeline. To subscribe to a subject, you would call `udpardRxSubscribtionInit` and then create a socket for this subscription. The socket and the subscription will exist side-by-side.

2. Set up static analysis and verification:
 - Clang-Tidy and SonarCloud with a large subset of MISRA rules enabled.
 - Build for i386, AMD64, ARM, AVR.
 - Run tests on i386 and AMD64 using both GCC and Clang with coverage.

3. Switch from Catch2 to Google Test.

### CODE REVIEW

File: libudpard/udpard.h [hunk 1]
```diff
@@ -412,13 +321,32 @@ UdpardTxQueue udpardTxInit(const size_t capacity, const size_t mtu_bytes);
...
+                        const UdpardConstPayload payload);
+
+int32_t udpardTxRespond(UdpardTx* const          self,
+                        void* const              user_transfer_reference,
+                        const UdpardMicrosecond  deadline_usec,
+                        const UdpardPriority     priority,
+                        const UdpardPortID       service_id,
+                        const UdpardNodeID       client_node_id,
+                        const UdpardTransferID   transfer_id,
+                        const UdpardConstPayload payload);
```
Comment (lydia-at-amazon) [inline libudpard/udpard.h:349]: What is the benefit of splitting this out into 3 separate functions? Only difference between the functions is the `UdpardNodeID`. We could call the parameter `remote_node_id` and just leave it unset for messages.

File: libudpard/udpard.h [hunk 2]
```diff
@@ -453,152 +381,167 @@ const UdpardTxQueueItem* udpardTxPeek(const UdpardTxQueue* const que);
...
+} UdpardRxServiceTransfer;
+
+int8_t udpardRxServiceDispatcherInit(UdpardRxServiceDispatcher* const self,
+                                     const UdpardNodeID               local_node_id,
+                                     const UdpardMemoryResource       memory_for_sessions,
+                                     const UdpardMemoryResource       memory_for_payloads);
+
+void udpardRxServiceDispatcherDestroy(UdpardRxServiceDispatcher* const self);
+
+int8_t udpardRxServiceDispatcherListen(UdpardRxServiceDispatcher* const self,
```
Comment (lydia-at-amazon) [inline libudpard/udpard.h:527]: What does this function do? How will it be used?

File: libudpard/udpard.h [hunk 3]
```diff
@@ -453,152 +381,167 @@ const UdpardTxQueueItem* udpardTxPeek(const UdpardTxQueue* const que);
...
+    /// Same applies to RPC-services as well.
+    ///
+    /// Once a session is created, it is never freed again until the port that owns it (this structure) is destroyed.
+    /// This is in line with the assumption that the network configuration is usually static, and that
+    /// once a node has started emitting data on a certain port, it is likely to continue doing so.
+    /// Applications where this is not the case may consider cycling their ports periodically
+    /// by destroying and re-creating them immediately.
+    ///
+    /// Each session instance takes sizeof(UdpardInternalRxSession) bytes of heap memory,
+    /// which is at most 128 bytes on most platforms (on small word size platforms it may be much smaller).
```
Comment (lydia-at-amazon) [inline libudpard/udpard.h:422]: We might need to provide a way to set the max number of sessions for platforms that have very limited memory.

File: libudpard/udpard.h (hunk 1 above)
Comment (pavel-kirienko) [reply libudpard/udpard.h:349]: That is not the only difference though. Notice that they accept the transfer-ID differently: the publish and request functions take it by reference and increment it, while the respond function does not need to modify the transfer-ID value as it is dictated by the client. We can unify them back if needed, but my assumption was that this design is clearer.

File: libudpard/udpard.h [hunk 4]
```diff
@@ -6,13 +6,29 @@
...
 ///
 /// description tbd
-/// This library is based heavily on libcanard
-// --------------------------------------------------------------------------------------------------------------------
+///
+/// Two decoupled parts of the library: TX pipeline and RX pipeline.
+/// The following sockets (or similar abstractions) are needed to use the library:
+///
+/// - One unconnected and unbound transmission socket used for all outgoing transfers:
+///   publications, RPC-requests, and RPC-responses.
```
[the remark this answers is not in the document]
Comment (pavel-kirienko) [reply libudpard/udpard.h]: Please see if the docs answer these questions.

File: libudpard/udpard.h [hunk 5]
```diff
@@ -453,152 +381,167 @@ const UdpardTxQueueItem* udpardTxPeek(const UdpardTxQueue* const que);
...
+    /// Each session instance takes sizeof(UdpardInternalRxSession) bytes of heap memory,
+    /// which is at most 128 bytes on most platforms (on small word size platforms it may be much smaller).
+    /// On top of that, each session instance may have one payload buffer allocated for the reassembled transfer;
+    /// the size of that buffer is determined by the application at the time of subscription.
+    /// The payload buffer is only allocated while reassembly is in progress; when the subscription is idle,
+    /// no additional memory is held by the session instance.
+    ///
+    /// From the above one can deduce that the worst situation from the memory management standpoint is when
+    /// there is a large number of nodes emitting data under a certain session specifier (e.g., publishing on a subject)
+    /// such that each node begins a multi-frame transfer while never completing it.
```
[the remark this answers is not in the document]
Comment (pavel-kirienko) [reply libudpard/udpard.h]: A typical application should never restart its sessions manually, the user needs not even know about it. The reassembly state machine will take care of all required state transitions automatically. One prospect for optimization is available when a node goes offline for good, rendering its session state unnecessary. Eventually, we can provide a dedicated API call to let the library know that a node such-and-such is no longer around and the memory help for its sessions can be released.

File: libudpard/udpard.h (hunk 3 above)
Comment (pavel-kirienko) [reply libudpard/udpard.h]: - How should the library respond if the limit is reached and a new session is needed to accept a transfer?
- Can we implement this via `UdpardMemoryResource`?

File: libudpard/udpard.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (lydia-at-amazon) [reply libudpard/udpard.h]: Sorry, I didn't mean make the user restart its sessions. I meant the library should restart the session. I like your idea of eventually providing a dedicated API to let the library know a node went permanently offline.

File: libudpard/udpard.h (hunk 1 above)
Comment (lydia-at-amazon) [reply libudpard/udpard.h]: Agreed that it is more clear to keep them separate.

File: libudpard/udpard.h [hunk 6]
```diff
@@ -563,42 +673,216 @@ int8_t udpardRxAccept(UdpardInstance* const         ins,
...
-/// The time complexity is logarithmic from the number of current subscriptions under the specified transfer kind.
+/// The time complexity is constant. This function does not invoke the dynamic memory manager.
+int8_t udpardRxServiceDispatcherInit(UdpardRxServiceDispatcher* const self,
+                                     const UdpardNodeID               local_node_id,
+                                     const UdpardMemoryResource       memory_for_sessions,
+                                     const UdpardMemoryResource       memory_for_payloads);
+
+/// Frees all memory held by the RPC-service dispatcher instance.
+/// After invoking this function, the instance is no longer usable.
+/// Do not forget to close the sockets that were opened for this instance.
```
Comment (thirtytwobits) [inline libudpard/udpard.h:832]: "...and, you know, maybe call your mother sometime."

File: libudpard/udpard.h [hunk 7]
```diff
@@ -563,42 +673,216 @@ int8_t udpardRxAccept(UdpardInstance* const         ins,
...
+    UdpardPortID     service_id;
+    bool             is_request;
+} UdpardRxServiceTransfer;
+
+/// To begin receiving RPC-service requests and/or responses, the application should do this:
+///
+///     1. Create a new UdpardRxServiceDispatcher instance.
+///
+///     2. Initialize it by calling udpardRxServiceDispatcherInit. Observe that a valid node-ID is required here.
+///        If the application has to perform a plug-and-play node-ID allocation, it has to complete that beforehand.
```
Comment (thirtytwobits) [inline libudpard/udpard.h:807]: Don't you need a dispatcher to participate in plug and play?

**Review by thirtytwobits — APPROVED** [review]
General comments. Looks good otherwise.

File: libudpard/udpard.h [hunk 8]
```diff
@@ -453,152 +381,167 @@ const UdpardTxQueueItem* udpardTxPeek(const UdpardTxQueue* const que);
...
+/// In Cyphal/UDP, each subject (topic) has a specific IP multicast group address associated with it.
+/// This address is available here in a dedicated field.
+/// The application is expected to open a socket bound to that endpoint and then feed the UDP datagrams received
+/// from that socket into the library by calling udpardRxSubscriptionReceive.
+typedef struct
+{
+    /// The IP address and UDP port number where UDP/IP datagrams matching this Cyphal subject will be sent.
+    /// The application should initialize a multicast socket bound to this endpoint.
+    /// READ-ONLY FIELD
+    UdpardUDPIPEndpoint udpip_endpoint;
```
[the remark this answers is not in the document]
Comment (maksimdrachov) [reply libudpard/udpard.h]: As a sidenote: Pavel is following [this](https://kb.zubax.com/pages/viewpage.action?pageId=2195699#ZubaxC++codingconventions-Types) C++ coding convention which I've found to be quite sensible.

(I can agree with `udp_ip_endpoint` being a bit more clear, `udp-ip_endpoint` does not roll quite as easy of the tongue/keyboard)

**Comment by pavel-kirienko** [comment]
@lydia-at-amazon @thirtytwobits I updated the docs to try and capture almost all of the issues raised here, and marked the respective threads as resolved.

File: libudpard/udpard.h [hunk 9]
```diff
@@ -28,67 +173,61 @@ extern "C" {
...
+/// This value represents an undefined node-ID: broadcast destination or anonymous source.
+#define UDPARD_NODE_ID_UNSET 0xFFFFU
 
 /// This is the recommended transfer-ID timeout value given in the Cyphal Specification. The application may choose
 /// different values per subscription (i.e., per data specifier) depending on its timing requirements.
-#define UDPARD_DEFAULT_TRANSFER_ID_TIMEOUT_USEC 2000000UL  /// 2 seconds
+#define UDPARD_DEFAULT_TRANSFER_ID_TIMEOUT_USEC 2000000UL
+
+/// The library supports at most this many redundant network interfaces per Cyphal node.
+#define UDPARD_NETWORK_INTERFACE_COUNT_MAX 3U
```
[the remark this answers is not in the document]
Comment (pavel-kirienko) [reply libudpard/udpard.h:223]: This is not a configuration parameter but a hard limit. What the user doesn't know is that this constant will define the number of RX reassembler states maintained per session within the library (this implementation detail is not visible at the API level).

image

**Comment by pavel-kirienko** [comment]
As this is approved, I am going to merge this now. Please move unfinished discussions to new issues if necessary.

File: libudpard/udpard.h (hunk 3 above)
Comment (lydia-at-amazon) [reply libudpard/udpard.h]: I think this was a more of a problem when libdupard was allocating 65535 RX session pointers in a UdpardRxSubscription: https://github.com/OpenCyphal-Garage/libudpard/blob/main/libudpard/udpard.h#L285 

Looks like your proposal removes this field, so we should be ok. Will resolve this comment.

Comment (pavel-kirienko) [reply libudpard/udpard.h]: Yes indeed, this is already addressed here.

**Pull request merged**